pyomo/contrib/pynumero/src/hsl_interface/ma57Interface.c

```diff
@@ -0,0 +1,343 @@
+		double* A) {
+
+	// Get memory estimates from INFO, allocate FACT and IFACT
+	if (!ma57->FACT_allocated) {
+		int info9 = ma57->INFO[9-1];
+		int size = (int)(ma57->FACT_factor*info9);
+		alloc_fact(ma57, size);
+	}
+	if (!ma57->IFACT_allocated) {
+		int info10 = ma57->INFO[10-1];
+    int size = (int)(ma57->IFACT_factor*info10);
+		alloc_ifact(ma57, size);
+	}
+
+	// Again, length of IWORK is a hard requirement
+	ma57->IWORK = malloc(N*sizeof(int));
+	if (ma57->IWORK == NULL) { abort_bad_memory(1); }
+
+	ma57bd_(&N, &NE, A, 
+			ma57->FACT, &(ma57->LFACT),
+			ma57->IFACT, &(ma57->LIFACT),
+			&(ma57->LKEEP), ma57->KEEP,
+			ma57->IWORK, ma57->ICNTL,
+			ma57->CNTL, ma57->INFO,
+			ma57->RINFO);
+
+	free(ma57->IWORK);
+}
+
+void do_backsolve(struct MA57_struct* ma57, int N, double* RHS) {
+
+	// Set number and length (principal axis) of RHS if not already set
+	if (!ma57->NRHS_set) {
+		set_nrhs(ma57, 1);
+	}
+	if (!ma57->LRHS_set) {
+		set_lrhs(ma57, N);
+	}
+
+	// Set JOB. Default is to perform full factorization
+	if (!ma57->JOB_set) {
+		set_job(ma57, 1);
+	}
+
+	// Allocate WORK if not done. Should be >= N
+	if (!ma57->WORK_allocated) {
+		int size = (int)(ma57->WORK_factor*ma57->NRHS*N);
+		alloc_work(ma57, size);
+	}
+
+	// IWORK should always be length N
+	ma57->IWORK = malloc(N*sizeof(int));
+	if (ma57->IWORK == NULL) { abort_bad_memory(1); }
+  
+	ma57cd_(
+			&(ma57->JOB), 
+			&N, 
+			ma57->FACT, 
+			&(ma57->LFACT),
+			ma57->IFACT, 
+			&(ma57->LIFACT),
+			&(ma57->NRHS), 
+			RHS,
+			&(ma57->LRHS), 
+			ma57->WORK,
+			&(ma57->LWORK), 
+			ma57->IWORK,
+			ma57->ICNTL, 
+			ma57->INFO
+			);
+
+	free(ma57->IWORK);
+	free(ma57->WORK);
+	ma57->WORK_allocated = false;
+}
+
+void do_iterative_refinement(struct MA57_struct* ma57, int N, int NE,
+		double* A, int* IRN, int* JCN, double* RHS, double* X, double* RESID) {
+	// Number of steps of iterative refinement can be controlled with ICNTL[9-1]
+
+	// Set JOB if not set. Controls how (whether) X and RESID will be used
+	if (!ma57->JOB_set) {
+		set_job(ma57, 1);
+	}
+
+	// Need to allocate WORK differently depending on ICNTL options
+	if (!ma57->WORK_allocated) {
+		int icntl9 = ma57->ICNTL[9-1];
+		int icntl10 = ma57->ICNTL[10-1];
+		int size;
+		if (icntl9 == 1) {
+			size = (int)(ma57->WORK_factor*N);
+		} else if (icntl9 > 1 && icntl10 == 0) {
+			size = (int)(ma57->WORK_factor*3*N);
+		} else if (icntl9 > 1 && icntl10 > 0) {
+			size = (int)(ma57->WORK_factor*4*N);
+		}
+		alloc_work(ma57, size);
+	}
+
+	ma57->IWORK = malloc(N*sizeof(int));
+	if (ma57->IWORK == NULL) { abort_bad_memory(1); }
+
+	ma57dd_(
+			&(ma57->JOB),
+			&N,
+			&NE,
+			IRN,
+			JCN,
+			ma57->FACT,
+			&(ma57->LFACT),
+			ma57->IFACT,
+			&(ma57->LIFACT),
+			RHS,
+			X,
+			RESID,
+			ma57->WORK,
+			ma57->IWORK,
+			ma57->ICNTL,
+			ma57->CNTL,
+			ma57->INFO,
+			ma57->RINFO
+			);
+
+	free(ma57->IWORK);
+	free(ma57->WORK);
+	ma57->WORK_allocated = false;
```

Comment:
My guess is ma27 does not have a subroutine like this one?

Reply:
Correct